matlab/doc/guideline_for_writing_tests_for_MATLAB_interface_for_apache_arrow.md

```diff
@@ -0,0 +1,97 @@
+<!---
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing,
+  software distributed under the License is distributed on an
+  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  KIND, either express or implied.  See the License for the
+  specific language governing permissions and limitations
+  under the License.
+-->
+
+# Guideline for Writing Tests for MATLAB Interface for Apache Arrow
+
+## Overview  
+
+This document is aimed at providing guidelines on testing of the `matlab` directory in the [`apache/arrow`](https://github.com/apache/arrow) repository.  
+
+## General Guideline  
+
+When designing any test case, please try to design it at the MATLAB interface level such that the tests are testing real use cases. Tests for MATLAB interface should use the MATLAB testing framework [`matlab.unittest.TestCase class`](https://www.mathworks.com/help/matlab/ref/matlab.unittest.testcase-class.html).
+
+If the source code cannot be covered by any test at the MATLAB interface level, consider adding c++ unit tests by using GoogleTest framework.
+
+## Code Coverage Goals  
+
+When making change to the `matlab` directory in the [`apache/arrow`](https://github.com/apache/arrow) repository, please try to add tests to cover all changed lines, conditions and decisions. It is understandable that some code cannot be hit due to defensive purpose or other reasons. When reviewer reviews code change, they'll check the code coverage for the changed code.  
+
+### How to Check Code Coverage  
+
+The Name-Value input argument [`ReportCoverageFor`](https://www.mathworks.com/help/matlab/ref/runtests.html#mw_764c9db7-6823-439f-a77d-7fd25a03d20e) to `runtests` is used to generate code coverage report.  
+
+
+```matlab  
+>> addpath( <your local arrow/matlab> )
+>> runtests(testFilePath/testFolderPath, 'ReportCoverageFor', sourceFilePath/sourceFolderPath, 'IncludeSubfolders', true/false);  
+
+% Example
+>> runtests('C:\TryCodeCoverage\arrow\matlab\test', 'ReportCoverageFor', 'C:\TryCodeCoverage\arrow\matlab\src\matlab\', 'IncludeSubfolders', true);
+```
+
+It might happen that caching or shadowing issue causes zero coverage because the test ends up executing some other source file. To avoid such problem, please first set a breakpoint in your source file and run the tests. This step is to make sure that your source file is executed by the tests. After this step, the command `runtests(___, "ReportCoverageFor", ___)` will work well.
```

Comment:
I reword the sentence as It might happen that caching or other issue causes zero coverage because... Kevin, you are right that this doesn't happen often. Move the paragraph to the 'Tips' section at the bottom of the document.